packages/bun-usockets/src/socket.c

```diff
@@ -363,54 +367,77 @@ int us_socket_write(int ssl, struct us_socket_t *s, const char *data, int length
         return us_internal_ssl_socket_write((struct us_internal_ssl_socket_t *) s, data, length);
     }
 #endif
-    if (us_socket_is_closed(ssl, s) || us_socket_is_shut_down(ssl, s)) {
+    if (us_socket_is_closed(ssl, s) || us_socket_is_shut_down(ssl, s) || !s->flags.is_writable) {
         return 0;
     }
 
-    int written = bsd_send(us_poll_fd(&s->p), data, length);
-    if (written != length) {
-        s->context->loop->data.last_write_failed = 1;
-        us_poll_change(&s->p, s->context->loop, LIBUS_SOCKET_READABLE | LIBUS_SOCKET_WRITABLE);
-    }
+    int total_written = 0;
+    int remaining = length;
+    do {
```

Comment:
What's the rationale behind a do-while loop vs. `while (remaining > 0) { ... }`?